### PR TITLE
Adds date range based model request endpoint

### PR DESCRIPTION
**Pull request by conbrad**
- Enables requesting model data based on station list and date range
# Test Links:
[Landing Page](https://wps-pr-2651.apps.silver.devops.gov.bc.ca/)
[MoreCast 2.0](https://wps-pr-2651.apps.silver.devops.gov.bc.ca/morecast-2)
[Percentile Calculator](https://wps-pr-2651.apps.silver.devops.gov.bc.ca/percentile-calculator)
[MoreCast](https://wps-pr-2651.apps.silver.devops.gov.bc.ca/morecast)
[C-Haines](https://wps-pr-2651.apps.silver.devops.gov.bc.ca/c-haines)
[FireBat](https://wps-pr-2651.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator)
[FireBat bookmark](https://wps-pr-2651.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN,s=9999&f=c7&c=NaN&w=NaN)
[Auto Spatial Advisory (ASA)](https://wps-pr-2651.apps.silver.devops.gov.bc.ca/auto-spatial-advisory)
[HFI Calculator](https://wps-pr-2651.apps.silver.devops.gov.bc.ca/hfi-calculator)

### CODE REVIEW

File: api/app/db/crud/weather_models.py [hunk 1]
```diff
@@ -241,6 +241,33 @@ def get_station_model_predictions(
...
+        order_by(WeatherStationModelPrediction.prediction_timestamp).\
+        group_by(WeatherStationModelPrediction.station_code, func.date(WeatherStationModelPrediction.prediction_timestamp)).\
+        having(
+            WeatherStationModelPrediction.prediction_timestamp == session.query(
+                func.max(WeatherStationModelPrediction.prediction_timestamp))
+            .filter(func.date(WeatherStationModelPrediction.prediction_timestamp) == func.date(WeatherStationModelPrediction.prediction_timestamp))
+            .filter(WeatherStationModelPrediction.station_code == WeatherStationModelPrediction.station_code)
+            .subquery()
+            .c.prediction_timestamp)
+    return query
```
Comment (dgboss) [inline api/app/db/crud/weather_models.py:268]: Does this return a query that is ready to be executed? Or is the query already executed and this is a list of results/rows?

Comment (conbrad) [reply api/app/db/crud/weather_models.py:268]: It returns a query that can be iterated over to grab each row. I'm not sure how it works under the hood I think it does sort of cursor based iterator.

**Review by dgboss — APPROVED** [review]

File: api/app/routers/weather_models.py [hunk 2]
```diff
@@ -43,14 +44,30 @@ async def get_most_recent_model_values(
...
-            stations=station_predictions)
-    except Exception as exception:
-        logger.critical(exception, exc_info=True)
-        raise
+    return WeatherStationsModelRunsPredictionsResponse(
+        stations=station_predictions)
+
+
+@router.post('/{model}/predictions/most_recent/{start_date}/{end_date}',
+             response_model=List[WeatherStationModelPredictionValues])
```
Comment (dgboss) [inline api/app/routers/weather_models.py:57]: Should we follow the general pattern of our other endpoints and explicitly define a response object that wraps the list? I think it's generally best practice to return an object/dictionary versus a list.

File: api/app/db/crud/weather_models.py [hunk 3]
```diff
@@ -241,6 +241,82 @@ def get_station_model_predictions(
...
+                            prediction_model_run_timestamps.id AS latest_timestamp_id,
+                            max(prediction_run_timestamp) AS latest_timestamp,
+                            date(prediction_run_timestamp) AS day,
+                            weather_station_model_predictions.station_code
+                        FROM
+                            prediction_model_run_timestamps,
+                            weather_station_model_predictions
+                        WHERE
+                            prediction_model_id IN (SELECT id FROM prediction_models WHERE abbreviation = '{model}')
+                            AND station_code IN ({",".join(str(s) for s in station_codes)})
```
Comment (andrea-williams) [inline api/app/db/crud/weather_models.py:276]: I'd prefer to see this translated into Python/sqlalchemy. That way we automatically have protection against the most common SQL injection attack methods.

Comment (conbrad) [reply api/app/db/crud/weather_models.py:276]: We already implicitly have this since these parameters are validated by pydantic

Comment (andrea-williams) [reply api/app/db/crud/weather_models.py:276]: pydantic doesn't know any more than to expect `station_codes` to be a list. 
`station_codes = ['1); DROP table;']` will satisfy pydantic, and will also cause trouble.

Comment (conbrad) [reply api/app/db/crud/weather_models.py:276]: Not true, it knows it should be a list of integers so that query is not possible.

Comment (andrea-williams) [reply api/app/db/crud/weather_models.py:276]: where is the list of ints specified? I only see a generic List in the function signature

Comment (conbrad) [reply api/app/db/crud/weather_models.py:276]: This is what I get back when I try that request body:

```
{
  "detail": [
    {
      "loc": [
        "body",
        "stations",
        0
      ],
      "msg": "value is not a valid integer",
      "type": "type_error.integer"
    }
  ]
}
```

Comment (conbrad) [reply api/app/db/crud/weather_models.py:276]: > where is the list of ints specified? I only see a generic List in the function signature
In the schema definitions:

```
class ModelDataRequest(BaseModel):
    """ A request for weather data for a given set of stations with a time of interest. """
    stations: List[int]
```

I can adjust the typings of the functions to use the specified type too for clarity

**Review by andrea-williams — APPROVED** [review]

File: api/app/routers/weather_models.py (hunk 2 above)
Comment (dgboss) [reply api/app/routers/weather_models.py:57]: Though it is unlikely anything else will ever consume this endpoint, so list versus dictionary probably doesn't matter.

**Pull request merged**